pkg/sync/originatorStream.go

```diff
@@ -0,0 +1,307 @@
+package sync
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/cenkalti/backoff/v5"
+	"github.com/xmtp/xmtpd/pkg/currency"
+	"github.com/xmtp/xmtpd/pkg/db"
+	"github.com/xmtp/xmtpd/pkg/db/queries"
+	envUtils "github.com/xmtp/xmtpd/pkg/envelopes"
+	"github.com/xmtp/xmtpd/pkg/fees"
+	"github.com/xmtp/xmtpd/pkg/metrics"
+	"github.com/xmtp/xmtpd/pkg/proto/xmtpv4/envelopes"
+	"github.com/xmtp/xmtpd/pkg/proto/xmtpv4/message_api"
+	"github.com/xmtp/xmtpd/pkg/registry"
+	"github.com/xmtp/xmtpd/pkg/utils"
+	"go.uber.org/zap"
+)
+
+type originatorStream struct {
+	ctx           context.Context
+	db            *sql.DB
+	log           *zap.Logger
+	node          *registry.Node
+	queries       *queries.Queries
+	lastEnvelope  *envUtils.OriginatorEnvelope
+	stream        message_api.ReplicationApi_SubscribeEnvelopesClient
+	feeCalculator fees.IFeeCalculator
+}
+
+func newOriginatorStream(
+	ctx context.Context,
+	db *sql.DB,
+	log *zap.Logger,
+	node *registry.Node,
+	lastEnvelope *envUtils.OriginatorEnvelope,
+	stream message_api.ReplicationApi_SubscribeEnvelopesClient,
+	feeCalculator fees.IFeeCalculator,
+) *originatorStream {
+	return &originatorStream{
+		ctx: ctx,
+		db:  db,
+		log: log.With(
+			zap.Uint32("originator_id", node.NodeID),
+			zap.String("http_address", node.HttpAddress),
+		),
+		node:          node,
+		queries:       queries.New(db),
+		lastEnvelope:  lastEnvelope,
+		stream:        stream,
+		feeCalculator: feeCalculator,
+	}
+}
+
+func (s *originatorStream) listen() error {
+	recvChan := make(chan *message_api.SubscribeEnvelopesResponse)
+	errChan := make(chan error)
+	writeQueue := make(chan *envUtils.OriginatorEnvelope, 10)
+
+	go func() {
+		for {
+			envs, err := s.stream.Recv()
+			if err != nil {
+				errChan <- err
+				return
+			}
+			recvChan <- envs
+		}
+	}()
+
+	go func() {
+		for {
+			select {
+			case <-s.ctx.Done():
+				return
+			case env := <-writeQueue:
+				if env == nil {
+					continue
+				}
+
+			storeLoop:
+				for {
+					err := s.storeEnvelope(env)
+					if err != nil {
+						s.log.Error("error storing envelope", zap.Error(err))
+						time.Sleep(1 * time.Second)
+						continue
```

Comment:
this might have to check for ctx.Done as well

Reply:
Good catch